File: backend/python-worker/tests/test_detection_taxonomy.py
"""Cross-runtime parity checks for the shared detection taxonomy."""
from __future__ import annotations

import json
import math
import sys
import unittest
from copy import deepcopy
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from detection.taxonomy import (
    DETECTION_TAXONOMY,
    ActiveModelInput,
    DetectionInputValidationError,
    DetectionTaxonomyValidationError,
    RegistryLabelInput,
    decode_detection_taxonomy,
    parse_active_model_input,
    parse_registry_label_input,
    resolve_label_capability,
)


class DetectionTaxonomyTests(unittest.TestCase):
    def test_unified_manifest_owns_coco_and_custom_classes_exactly(self) -> None:
        active_model = {
            "versionKey": "unified-v1",
            "runtimeMode": "UNIFIED",
            "labelMap": {"truck": "truck", "reach_stacker": "reach_stacker"},
        }
        truck = resolve_label_capability(
            {"vietnameseName": "Xe táº£i", "baseClass": "truck"},
            active_model,
        )
        missing_car = resolve_label_capability(
            {"vietnameseName": "Xe con", "baseClass": "car"},
            active_model,
        )
        self.assertEqual(truck.detection_source, "CUSTOM")
        self.assertEqual(truck.reason_code, "ACTIVE_UNIFIED_CLASS")
        self.assertEqual(truck.active_model_version, "unified-v1")
        self.assertFalse(missing_car.is_detectable)
        self.assertEqual(missing_car.reason_code, "UNIFIED_CLASS_NOT_IN_ACTIVE_MODEL")

    def test_legacy_manifest_defaults_to_supplemental_and_runtime_mode_is_strict(self) -> None:
        parsed = parse_active_model_input({
            "versionKey": "legacy-v1",
            "labelMap": {"reach": "reach_stacker"},
        })
        self.assertEqual(parsed.runtime_mode, "SUPPLEMENTAL")
        with self.assertRaises(DetectionInputValidationError):
            parse_active_model_input({
                "versionKey": "bad-v1",
                "runtimeMode": "PRIMARY",
                "labelMap": {"reach": "reach_stacker"},
            })

    def test_exact_active_manifest_repairs_legacy_xe_nang_read_routing(self) -> None:
        capability = resolve_label_capability(
            {"vietnameseName": "Xe nâng", "baseClass": "truck"},
            {"versionKey": "custom-legacy", "labelMap": {"Xe nâng": "reach stacker"}},
        )

        self.assertEqual(capability.canonical_class, "reach_stacker")
        self.assertEqual(capability.detection_source, "CUSTOM")
        self.assertTrue(capability.is_detectable)
        self.assertEqual(capability.reason_code, "ACTIVE_CUSTOM_LEGACY_LABEL")

    @classmethod
    def setUpClass(cls) -> None:
        cases_path = Path(__file__).resolve().parents[2] / "config" / "detection-taxonomy-cases.json"
        cls.fixture = json.loads(cases_path.read_text(encoding="utf-8"))
        taxonomy_path = Path(__file__).resolve().parents[2] / "config" / "detection-taxonomy.json"
        cls.raw_taxonomy = json.loads(taxonomy_path.read_text(encoding="utf-8"))

    def test_shared_parity_cases(self) -> None:
        self.assertEqual(self.fixture["schemaVersion"], 1)
        self.assertEqual(len(self.fixture["cases"]), 19)  # original 13, reserved-name parity, shared whitespace, U+001C edges
        for test_case in self.fixture["cases"]:
            with self.subTest(test_case["name"]):
                actual = resolve_label_capability(test_case["label"], test_case["activeModel"]).as_dict()
                self.assertEqual(actual, test_case["expected"])

    def test_sample_count_does_not_affect_runtime_capability(self) -> None:
        zero_samples = {"vietnameseName": "Xe nâng container", "baseClass": "reach_stacker", "sampleCount": 0}
        many_samples = {"vietnameseName": "Xe nâng container", "baseClass": "reach_stacker", "sampleCount": 10_000}
        active_model = {"versionKey": "custom-v1", "labelMap": {"Xe nâng container": "reach_stacker"}}
        self.assertEqual(resolve_label_capability(zero_samples, active_model), resolve_label_capability(many_samples, active_model))

    def test_capability_ignores_malformed_or_direct_sample_count(self) -> None:
        active_model = {"versionKey": "custom-v1", "labelMap": {"Xe nâng container": "reach_stacker"}}
        expected = resolve_label_capability(
            {"vietnameseName": "Xe nâng container", "baseClass": "reach_stacker"},
            active_model,
        )
        self.assertEqual(
            resolve_label_capability(
                {
                    "vietnameseName": "Xe nâng container",
                    "baseClass": "reach_stacker",
                    "sampleCount": {"malformed": True},
                    "_count": {"samples": float("nan")},
                },
                active_model,
            ),
            expected,
        )
        self.assertEqual(
            resolve_label_capability(RegistryLabelInput("Xe tải", "truck", -1), None).reason_code,
            "COCO_BASE_CLASS",
        )

    def test_malformed_input_parity_fixture(self) -> None:
        for test_case in self.fixture["malformedInputs"]:
            parser = parse_registry_label_input if test_case["target"] == "registry" else parse_active_model_input
            with self.subTest(test_case["name"]):
                with self.assertRaises(DetectionInputValidationError) as raised:
                    parser(test_case["value"])
                self.assertEqual(raised.exception.category, test_case["expectedCategory"])

    def test_direct_dataclass_inputs_are_revalidated_at_resolver_boundary(self) -> None:
        invalid_cases = (
            (
                "blank direct ActiveModelInput version",
                RegistryLabelInput("X", "reach_stacker"),
                ActiveModelInput("   ", {"X": "reach_stacker"}),
            ),
            (
                "invalid direct ActiveModelInput labelMap",
                RegistryLabelInput("X", "reach_stacker"),
                ActiveModelInput("custom-v1", {"X": "container"}),
            ),
        )
        for name, label, active_model in invalid_cases:
            with self.subTest(name):
                with self.assertRaises(DetectionInputValidationError) as raised:
                    resolve_label_capability(label, active_model)
                self.assertEqual(raised.exception.category, "INPUT_VALIDATION")

    def test_malformed_taxonomy_parity_fixture(self) -> None:
        for test_case in self.fixture["malformedTaxonomy"]:
            candidate = deepcopy(self.raw_taxonomy)
            current = candidate
            for key in test_case["path"][:-1]:
                current = current[key]
            leaf = test_case["path"][-1]
            if test_case.get("delete"):
                del current[leaf]
            else:
                current[leaf] = test_case["value"]
            with self.subTest(test_case["name"]):
                with self.assertRaises(DetectionTaxonomyValidationError) as raised:
                    decode_detection_taxonomy(candidate)
                self.assertEqual(raised.exception.category, test_case["expectedCategory"])

    def test_schema_accepts_integral_json_numbers_and_normalizes_to_int(self) -> None:
        candidate = deepcopy(self.raw_taxonomy)
        candidate["schemaVersion"] = 1.0
        candidate["cocoClasses"]["person"] = 0.0
        normalized = decode_detection_taxonomy(candidate)
        self.assertEqual(normalized["schemaVersion"], 1)
        self.assertEqual(normalized["cocoClasses"]["person"], 0)
        self.assertIsInstance(normalized["schemaVersion"], int)
        self.assertIsInstance(normalized["cocoClasses"]["person"], int)

    def test_schema_rejects_nonintegral_or_unsafe_values(self) -> None:
        for value in (True, 1.5, math.nan, math.inf, -math.inf, 9_007_199_254_740_992):
            with self.subTest(value=value):
                candidate = deepcopy(self.raw_taxonomy)
                candidate["schemaVersion"] = value
                with self.assertRaises(DetectionTaxonomyValidationError):
                    decode_detection_taxonomy(candidate)

    def test_coco_ids_reject_boolean_fractional_nonfinite_or_unsafe_values(self) -> None:
        for value in (True, 0.5, math.nan, math.inf, -math.inf, 9_007_199_254_740_992):
            with self.subTest(value=value):
                candidate = deepcopy(self.raw_taxonomy)
                candidate["cocoClasses"]["person"] = value
                with self.assertRaises(DetectionTaxonomyValidationError):
                    decode_detection_taxonomy(candidate)

    def test_sample_count_accepts_integral_json_float_and_normalizes_to_int(self) -> None:
        parsed = parse_registry_label_input({
            "vietnameseName": "Xe tải",
            "baseClass": "truck",
            "sampleCount": 0.0,
        })
        self.assertEqual(parsed.sample_count, 0)
        self.assertIsInstance(parsed.sample_count, int)

    def test_taxonomy_is_deeply_immutable(self) -> None:
        with self.assertRaises(TypeError):
            DETECTION_TAXONOMY["cocoClasses"]["person"] = 79
        with self.assertRaises(TypeError):
            DETECTION_TAXONOMY["legacyNameConstraints"]["xe nâng"] = ("truck",)
        with self.assertRaises(AttributeError):
            DETECTION_TAXONOMY["legacyNameConstraints"]["xe nâng"].append("truck")


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: backend/node-api/src/tests/test_detection_taxonomy.ts
import assert from 'assert';
import fs from 'fs';
import path from 'path';

import {
  decodeDetectionTaxonomy,
  DETECTION_TAXONOMY,
  DetectionCapability,
  DetectionInputValidationError,
  DetectionTaxonomyValidationError,
  parseActiveModelInput,
  parseRegistryLabelInput,
  resolveLabelCapability,
} from '../detection/taxonomy';

const unifiedModel = {
  versionKey: 'unified-v1',
  runtimeMode: 'UNIFIED',
  labelMap: { truck: 'truck', reach_stacker: 'reach_stacker' },
};
assert.deepStrictEqual(
  resolveLabelCapability({ vietnameseName: 'Xe tải', baseClass: 'truck' }, unifiedModel),
  {
    canonicalClass: 'truck',
    detectionSource: 'CUSTOM',
    isDetectable: true,
    activeModelVersion: 'unified-v1',
    reasonCode: 'ACTIVE_UNIFIED_CLASS',
    reasonText: 'Nhận diện bởi model unified unified-v1',
  },
);
const missingUnifiedCar = resolveLabelCapability(
  { vietnameseName: 'Xe con', baseClass: 'car' },
  unifiedModel,
);
assert.equal(missingUnifiedCar.isDetectable, false);
assert.equal(missingUnifiedCar.reasonCode, 'UNIFIED_CLASS_NOT_IN_ACTIVE_MODEL');
assert.equal(
  parseActiveModelInput({ versionKey: 'legacy-v1', labelMap: { reach: 'reach_stacker' } }).runtimeMode,
  'SUPPLEMENTAL',
);
assert.throws(
  () => parseActiveModelInput({ versionKey: 'bad-v1', runtimeMode: 'PRIMARY', labelMap: { reach: 'reach_stacker' } }),
  DetectionInputValidationError,
);

assert.deepStrictEqual(
  resolveLabelCapability(
    { vietnameseName: 'Xe nâng', baseClass: 'truck' },
    { versionKey: 'custom-legacy', labelMap: { 'Xe nâng': 'reach stacker' } },
  ),
  {
    canonicalClass: 'reach_stacker',
    detectionSource: 'CUSTOM',
    isDetectable: true,
    activeModelVersion: 'custom-legacy',
    reasonCode: 'ACTIVE_CUSTOM_LEGACY_LABEL',
    reasonText: 'Nhận diện bởi model custom custom-legacy; nhãn legacy được định nghĩa bởi manifest',
  },
);

interface TaxonomyCase {
  name: string;
  label: unknown;
  activeModel: unknown | null;
  expected: DetectionCapability;
}

interface MalformedInputCase {
  name: string;
  target: 'registry' | 'activeModel';
  value: unknown;
  expectedCategory: 'INPUT_VALIDATION';
}

interface MalformedTaxonomyCase {
  name: string;
  path: string[];
  value?: unknown;
  delete?: boolean;
  expectedCategory: 'TAXONOMY_VALIDATION';
}

interface TaxonomyCasesFile {
  schemaVersion: number;
  cases: TaxonomyCase[];
  malformedInputs: MalformedInputCase[];
  malformedTaxonomy: MalformedTaxonomyCase[];
}

const casesPath = path.resolve(__dirname, '../../../config/detection-taxonomy-cases.json');
const fixture = JSON.parse(fs.readFileSync(casesPath, 'utf8')) as TaxonomyCasesFile;

assert.equal(fixture.schemaVersion, 1);
assert.equal(fixture.cases.length, 19, 'the original 13 parity cases, reserved-name parity, shared-whitespace and all U+001C edge cases');

for (const testCase of fixture.cases) {
  assert.deepStrictEqual(
    resolveLabelCapability(testCase.label, testCase.activeModel),
    testCase.expected,
    testCase.name,
  );
}

const zeroSampleResult = resolveLabelCapability(
  { vietnameseName: 'Xe nâng container', baseClass: 'reach_stacker', sampleCount: 0 },
  { versionKey: 'custom-v1', labelMap: { 'Xe nâng container': 'reach_stacker' } },
);
const manySampleResult = resolveLabelCapability(
  { vietnameseName: 'Xe nâng container', baseClass: 'reach_stacker', sampleCount: 10_000 },
  { versionKey: 'custom-v1', labelMap: { 'Xe nâng container': 'reach_stacker' } },
);
assert.deepStrictEqual(manySampleResult, zeroSampleResult, 'sampleCount must not affect runtime capability');

const malformedFakeCountResult = resolveLabelCapability(
  {
    vietnameseName: 'Xe nâng container',
    baseClass: 'reach_stacker',
    sampleCount: { malformed: true },
    _count: { samples: Number.NaN },
  },
  { versionKey: 'custom-v1', labelMap: { 'Xe nâng container': 'reach_stacker' } },
);
assert.deepStrictEqual(
  malformedFakeCountResult,
  zeroSampleResult,
  'malformed DTO-only sample count fields must not affect runtime capability',
);

function assertInputCategory(action: () => unknown, name: string): void {
  assert.throws(action, (error: unknown) =>
    error instanceof DetectionInputValidationError && error.category === 'INPUT_VALIDATION', name);
}

for (const testCase of fixture.malformedInputs) {
  assertInputCategory(
    () => (testCase.target === 'registry' ? parseRegistryLabelInput(testCase.value) : parseActiveModelInput(testCase.value)),
    testCase.name,
  );
}

const taxonomyPath = path.resolve(__dirname, '../../../config/detection-taxonomy.json');
const rawTaxonomy = JSON.parse(fs.readFileSync(taxonomyPath, 'utf8')) as Record<string, unknown>;

function clonedTaxonomy(): Record<string, unknown> {
  return JSON.parse(JSON.stringify(rawTaxonomy)) as Record<string, unknown>;
}

function mutateAtPath(document: Record<string, unknown>, testCase: MalformedTaxonomyCase): void {
  let current: Record<string, unknown> = document;
  for (const key of testCase.path.slice(0, -1)) {
    const nested = current[key];
    assert.ok(nested !== null && typeof nested === 'object' && !Array.isArray(nested), testCase.name);
    current = nested as Record<string, unknown>;
  }
  const leaf = testCase.path[testCase.path.length - 1];
  if (testCase.delete === true) {
    delete current[leaf];
  } else {
    current[leaf] = testCase.value;
  }
}

for (const testCase of fixture.malformedTaxonomy) {
  const invalid = clonedTaxonomy();
  mutateAtPath(invalid, testCase);
  assert.throws(
    () => decodeDetectionTaxonomy(invalid),
    (error: unknown) => error instanceof DetectionTaxonomyValidationError
      && error.category === testCase.expectedCategory,
    testCase.name,
  );
}

// This explicit object-level check guards JavaScript's boolean coercion boundary.
const booleanSchema = clonedTaxonomy();
booleanSchema.schemaVersion = true;
assert.throws(() => decodeDetectionTaxonomy(booleanSchema), DetectionTaxonomyValidationError);

const integralJsonNumbers = clonedTaxonomy();
integralJsonNumbers.schemaVersion = 1.0;
(integralJsonNumbers.cocoClasses as Record<string, unknown>).person = 0.0;
const normalizedIntegralTaxonomy = decodeDetectionTaxonomy(integralJsonNumbers);
assert.equal(normalizedIntegralTaxonomy.schemaVersion, 1);
assert.equal(normalizedIntegralTaxonomy.cocoClasses.person, 0);

for (const value of [true, 1.5, Number.NaN, Number.POSITIVE_INFINITY, Number.NEGATIVE_INFINITY, Number.MAX_SAFE_INTEGER + 1]) {
  const invalid = clonedTaxonomy();
  invalid.schemaVersion = value;
  assert.throws(() => decodeDetectionTaxonomy(invalid), DetectionTaxonomyValidationError, `schemaVersion rejects ${String(value)}`);
}

for (const value of [true, 0.5, Number.NaN, Number.POSITIVE_INFINITY, Number.NEGATIVE_INFINITY, Number.MAX_SAFE_INTEGER + 1]) {
  const invalid = clonedTaxonomy();
  (invalid.cocoClasses as Record<string, unknown>).person = value;
  assert.throws(() => decodeDetectionTaxonomy(invalid), DetectionTaxonomyValidationError, `COCO ID rejects ${String(value)}`);
}

const normalizedSampleCount = parseRegistryLabelInput({
  vietnameseName: 'Xe tải',
  baseClass: 'truck',
  sampleCount: 0.0,
});
assert.equal(normalizedSampleCount.sampleCount, 0);

assert.ok(Object.isFrozen(DETECTION_TAXONOMY));
assert.ok(Object.isFrozen(DETECTION_TAXONOMY.cocoClasses));
assert.ok(Object.isFrozen(DETECTION_TAXONOMY.legacyNameConstraints));
assert.ok(Object.isFrozen(DETECTION_TAXONOMY.legacyNameConstraints['xe nâng']));
assert.throws(() => {
  (DETECTION_TAXONOMY.cocoClasses as Record<string, number>).person = 79;
}, TypeError);
assert.throws(() => {
  (DETECTION_TAXONOMY.legacyNameConstraints['xe nâng'] as string[]).push('truck');
}, TypeError);

console.log(`detection taxonomy: PASS (${fixture.cases.length} parity cases, ${fixture.malformedInputs.length + fixture.malformedTaxonomy.length} malformed cases)`);
